[mlir][Linalg] Allow more control in drop unit dims

sommerlukas opened the pull request:
Extend the `ControlDropUnitDims` struct to allow users of the `linalg::dropUnitDims` function more control over the behavior of the function.

The extended struct allows users to specify functions to:
- Calculate the shape and new index map of operands, which also allows to control which operands get their unit extent dimension dropped.
- How the operands are collapsed
- How the result is expanded to the original shape

One example (and the motivation for this change) where this additional control is useful is to allow collapsing of tensors with an encoding, as demonstrated by the new test.

The default implementations preserve the previous behavior; existing users of the interface do not need to make any changes to their code.

sommerlukas committed:
[mlir][Linalg] Allow more control in drop unit dims

Extend the `ControlDropUnitDims` struct to allow users of the `linalg::dropUnitDims` function more control over the behavior of the function.

The extended struct allows users to specify functions to:
- Calculate the shape and new index map of operands, which also allows to control which operands get their unit extent dimension dropped.
- How the operands are collapsed
- How the result is expanded to the original shape

One example (and the motivation for this change) where this additional control is useful is to preserve tensor encodings and allow collapsing of tensors with an encoding, as demonstrated by the new test.

The default implementations preserve the previous behavior, existing users of the interface do not need to make any changes to their code.

Signed-off-by: Lukas Sommer <lukas.sommer@amd.com>

rengolin reviewed (commented):
If preserving the encoding defaults to the previous behaviour, why isn't this the new default behaviour, instead of a test pass implementation?

sommerlukas committed:
Address PR feedback

Flip logic to enable simpler early exit.

Rename option in test pass to clarify behavior.

Signed-off-by: Lukas Sommer <lukas.sommer@amd.com>

banach-space reviewed (commented):
Thanks for contributing!

What should happen if tensors use different encodings? Also:

> The extended struct allows users to specify functions to:
> 
> * Calculate the shape and new index map of operands, which also allows to control which operands get their unit extent dimension dropped.
> * How the operands are collapsed
> * How the result is expanded to the original shape

That's 3 things, but only one new test is added.

Btw, would object to reducing the density of lambdas? Basically, I see quite a few nested functions that I find hard to parse and which, IMHO, should be documented. Perhaps this could be improved with comments alone 🤷🏻

rengolin commented:
> So, the default behavior has always been to preserve encoding and will continue to do so, but it doesn't collapse operands with encoding. The new controls enable to do that as one possible use case.

If the end result still returns the same shape (`<1, 1, 42>`) via collapse and expand shapes, then what's the benefit of this pass?

If you want to return some other shape, then can't we just collapse the resulting shape?

The example doesn't seem to convey the real motivation behind this extra control (and interface complexity).

sommerlukas committed:
Improve documentation and test

Add Doxzgen-style comments on the new function types and members of the control struct.

Overhaul the test to use a single input IR for both modes and test custom expansion by adding an attribute on the operation.

Signed-off-by: Lukas Sommer <lukas.sommer@amd.com>

sommerlukas commented on mlir/test/Dialect/Linalg/test-drop-unit-dims.mlir:
```diff
@@ -24,3 +25,44 @@ func.func @drop_outermost_unit_dims(%arg0: tensor<1x1x42xf32>) -> tensor<1x1x42x
+        linalg.yield %2 : f32
+    } -> tensor<1x1x42xf32, #encoding>
+  return %1 : tensor<1x1x42xf32, #encoding>
+}
```
I've overhauled the test to only use a single IR for both modes now.

In the default mode, the input with the encoding isn't collapsed and retains its original shape. 

In the new mode, that input is also collapsed and has shape `1x42xf32`.

I've also extended the new mode of the test pass to now cover all three new interfaces:
> Calculate the shape and new index map of operands, which also allows to control which operands get their unit extent dimension dropped.

The new mode uses a custom implementation to allow collapsing operands with encoding.

> How the operands are collapsed

The new mode uses a custom implementation to collapse the operand and retain the original encoding.

> How the result is expanded to the original shape

In the new mode, a discardable attribute is attached to the expand operation to demonstrate that the behavior can be customized.

sommerlukas commented:
The rationale to put the partial (or in case of `collapseValueFn` and `expandValueFn`, full) implementation into the `.cpp` file was to avoid "polluting" the header with that code. If you think it would avoid confusion, I can move the entire default implementations into the lambda instead.

I chose to use `std::function`/lambdas because the existing `controlFn` also uses a lambda and, in general, the use of `std::function` to customize behavior seems quite common in this header, so I tried to follow the existing style.

What you have in mind would be more along the lines of class inheritance and overriding, with dynamic (`virtual`) or static polymorphism?

MaheshRavishankar reviewed (changes requested):
Thanks @sommerlukas  for taking a stab at this. I should have probably provided a bit more context for you before sending you down this road :) . I think the options struct here is way to heavy-weight. The options struct should just be a container for the different ways in which the transformations can be controlled. I think here the options struct is now encompassing actual implementation details of the transformations as methods. The options struct should just have fields that represent the different options, and methods to set/retrieve those options. 

In this context, I think this is the options struct should contain the following. 
1. The current control function. The control function already allows to control which dimensions are being dropped. That drives the calculation of the reassociation indices needed for expand and collapse. The only issue is w.r.t how to handle operands/results that have encodings. If there are no encodings, the transformation already knows what to do. If there is an encoding, we should
2. The collapseValueFn callback should take as arguments the original operand and reassociation indices that tell you which dimensions of the original values to collapse. It should return the new value to use as operands to the newly created operation. In other words, without encoding the original value -> new value is obtained by using a `tensor.collapse_shape`.
3. The expandValueFn callback should take as arguments the new result and the reassociation indices that tell you which dimensions of the new result value need to be expanded to get back the original rank. It should return the new value to use as replacements for the original result. In other words, without encoding the new result -> original value replacement is obtained using a `tensor.expand_shape`.

I think the change would be much more simple and targeted. It will be a breaking change, but thats OK. Its a one-time breaking change to use the options struct, after which the API wont break, and will be strictly additive.

Happy to chat on a video call if you have more questions.
